File: leetcode/0213-rob.py
class Solution:
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums is None or len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        return max(self._rob(nums, 0, len(nums)-2), self._rob(nums, 1, len(nums)-1))
    
    def _rob(self, nums, low, high):
        cur, pre_one, pre_two = nums[low], nums[low], 0
        for num in nums[low+1:high+1]:
            cur = max(pre_one, pre_two+num)
            pre_one, pre_two = cur, pre_one
        return cur

File: leetcode/0213-rob.java
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0)
            return 0;
        int robFirst = nums[0], preRobFirst = 0;
        int notRobFirst = 0, preNotRobFirst = 0;
        int curRobFirst = nums[0], curNotRobFirst = 0;
        for (int i = 1; i < nums.length; ++i) {
            if (i != nums.length-1) {
                curRobFirst = Math.max(preRobFirst+nums[i], robFirst);
                curNotRobFirst = Math.max(preNotRobFirst+nums[i], notRobFirst);
                int tmp = robFirst;
                robFirst = curRobFirst;
                preRobFirst = tmp;
                tmp = notRobFirst;
                notRobFirst = curNotRobFirst;
                preNotRobFirst = tmp;
            } else {
                curNotRobFirst = Math.max(preNotRobFirst+nums[i], notRobFirst);
            }
        }
        return Math.max(curNotRobFirst, curRobFirst);        
    }
}
